mempool/src/shared_mempool/peer_manager.rs

```diff
@@ -1,121 +1,148 @@
+
+            if !peer_state.is_alive {
+                self.picked_fallback_peers.remove(&peer);
+            }
+
+            let broadcast_peers_count = alive_primaries + self.picked_fallback_peers.len();
+            if broadcast_peers_count < self.min_broadcast_recipient_count {
+                // pick more fallback peers
+                let new_fallback_peers: Vec<PeerNetworkId> = self
+                    .peer_info
+                    .iter()
+                    .filter(|(peer, state)| !self.is_picked_peer(**peer) && state.is_alive)
+                    .map(|(peer, _state)| *peer)
+                    .choose_multiple(
+                        &mut rand::thread_rng(),
+                        self.min_broadcast_recipient_count - broadcast_peers_count,
+                    )
+                    .into_iter()
+                    .collect();
+
+                self.picked_fallback_peers
+                    .extend(new_fallback_peers.clone());
+                new_picked_peers.extend(new_fallback_peers);
+            }
+        }
+
+        // schedule broadcast for newly picked peers
+        for peer in new_picked_peers {
+            if let Err(_e) = &self.schedule_broadcast_tx.push(peer, peer) {
+                error!("failed to schedule broadcast for new peer");
+            }
         }
-        picked_peers
     }
 
-    pub fn update_peer_broadcast(&self, peer_broadcasts: Vec<(PeerId, u64)>) {
-        let mut peer_info = self
-            .peer_info
-            .lock()
-            .expect("failed to acquire peer_info lock");
+    pub fn update_peer_broadcast(&mut self, update: (PeerNetworkId, u64)) {
+        let (peer, new_timeline_id) = update;
+        self.peer_info.entry(peer).and_modify(|t| {
+            t.timeline_id = new_timeline_id;
+        });
+    }
+
+    pub fn is_upstream_peer(&self, peer: PeerNetworkId) -> bool {
+        self.upstream_config.is_upstream_peer(peer)
+    }
+
+    fn is_primary_upstream_peer(&self, peer: PeerNetworkId) -> bool {
+        self.upstream_config.is_primary_upstream_peer(peer)
+    }
 
-        for (peer_id, new_timeline_id) in peer_broadcasts {
-            peer_info.entry(peer_id).and_modify(|t| {
-                t.timeline_id = new_timeline_id;
-            });
+    // returns a peer's sync state if the peer is a chosen broadcast peer
+    pub fn get_picked_state(&self, peer: PeerNetworkId) -> Option<PeerSyncState> {
+        if self.is_picked_peer(peer) {
+            self.peer_info.get(&peer).cloned()
+        } else {
+            None
         }
     }
 
-    pub fn is_upstream_peer(&self, network_id: PeerId, peer_id: PeerId) -> bool {
-        self.upstream_config.is_upstream_peer(network_id, peer_id)
+    // checks whether a peer is a chosen broadcast recipient:
+    // - all primary peers
+    // - select fallback peers, if k-policy is enabled
+    // this does NOT check for whether this peer is alive
+    fn is_picked_peer(&self, peer: PeerNetworkId) -> bool {
```

Comment:
qq: why won't try to always schedule broadcast for new_peer event (and actually even ignore lost peer event) and then check if peer is picked on a fly? 
so if it's not it's just noop routine
also you won't need to persist `picked_fallback_peers` in this case, but I'm not sure I fully understand its use